src/assets/TokensController.ts

```diff
@@ -268,6 +294,72 @@ export class TokensController extends BaseController<
     }
   }
 
+  /**
+   * Adds isERC721 field to token object
+   * (Called when a user attempts to add tokens that were previously added which do not yet had isERC721 field)
+   *
+   * @param {string} tokenAddress - The contract address of the token requiring the isERC721 field added.
+   * @returns The new token object with the added isERC721 field.
+   *
+   */
+  async updateTokenType(tokenAddress: string) {
+    const isERC721 = await this._detectIsERC721(tokenAddress);
+    const { tokens } = this.state;
+    const tokenIndex = tokens.findIndex((token) => {
+      return token.address === tokenAddress;
+    });
+    tokens[tokenIndex].isERC721 = isERC721;
+    this.update({ tokens });
+    return tokens[tokenIndex];
+  }
+
+  /**
+   * Detects whether or not a token is ERC-721 compatible.
+   *
+   * @param {string} tokensAddress - the token contract address.
+   * @returns boolean indicating whether the token address passed in supports the EIP-721 interface.
+   *
+   */
+  async _detectIsERC721(tokenAddress: string) {
+    const checksumAddress = toChecksumHexAddress(tokenAddress);
+    // if this token is already in our contract metadata map we don't need
+    // to check against the contract
+    if (contractsMap[checksumAddress]?.erc721 === true) {
+      return Promise.resolve(true);
+    }
+    const tokenContract = await this._createEthersContract(
+      tokenAddress,
+      abiERC721,
+      this.ethersProvider,
+    );
+
+    return await tokenContract
+      .supportsInterface(ERC721_INTERFACE_ID)
+      .catch((error: { code: string; method: string }) => {
+        if (error.code === 'UNPREDICTABLE_GAS_LIMIT') {
+          return false;
+        }
+        throw error;
+      });
```

Comment:
Nit: Using `await` rather than chaining Promises has better support for async stack traces in Chrome (and hopefully soon in Firefox).

```suggestion
    try {
      return await tokenContract
        .supportsInterface(ERC721_INTERFACE_ID);
    } catch (error: Error) {
        if (error.code === 'UNPREDICTABLE_GAS_LIMIT') {
          return false;
        }
        throw error;
    }
```

Also the error type should be `Error`. It has more than just those two properties, and various libraries like Sentry to treat it differently.

Reply:
Also, we should have a comment explaining that we don't know why `UNPREDICTABLE_GAS_LIMIT` is thrown here, but that it's what we see for ERC20 tokens. That could be confusing later!

Reply:
Ok. I originally had the type set to `Error` but it doesn't support code as a field?

Reply:
Actually it looks like we can't specify an error type in a try/catch in ts? That seems wrong...
![Screen Shot 2021-07-16 at 9 32 12 AM](https://user-images.githubusercontent.com/34557516/125964079-ec228630-5555-4f57-9083-a5760cfd8730.png)

Reply:
Huh, TIL. I'm now noticing that I haven't actually declared a type on a catch block before, it's always been inferred.

`unknown` probably would be best because you can throw anything really. Just that it's very inconvenient for debugging if anything but an `Error` is thrown. So I guess I was hoping declaring it as an `Error` would make more likely that TypeScript would attempt to make it so, but that's not how any of this works.

Reply:
Think it may have to be `any` :(
![Screen Shot 2021-07-16 at 9 58 05 AM](https://user-images.githubusercontent.com/34557516/125967773-f1cee71e-b9bd-47c3-8e7a-b49b4903c100.png)

Reply:
Made type `any` for now...

Reply:
That error is the TypeScript compiler asking you to further narrow the type. Something like `typeof error === 'object' &&` might do the trick! Though to be honest I would have expected the optional chain operator to work :thinking: Strange.

Reply:
Wow, I gave this a shot and it's really not making this easy :grimacing:  Not blocking on this, but I did think the whole point of the `unknown` type was to force you to not make assumptions about the type, and... I think we're doing that well here despite the compiler's complaints. So that's a shame.

Reply:
I made a thread about this in the dev channel. I think I found a solution but we can pursue this some other time.